tests/algorithms/rank_relevancy/test_RR_abstract.py

```diff
@@ -5,23 +5,25 @@
 
 from smqtk_relevancy.interfaces.rank_relevancy import RankRelevancyWithFeedback
 
+from typing import Dict, Any
+
 
 class DummyRRWF(RankRelevancyWithFeedback):
-    def __init__(self):
+    def __init__(self) -> None:
         self._rwf_mock = mock.Mock()
 
     # Implement RankRelevancyWithFeedback._rank_with_feedback using a
     # per-instance Mock.  self._rank_with_feedback is thus a Mock
     # instead of a bound method.
     @property
-    def _rank_with_feedback(self):
+    def _rank_with_feedback(self) -> "RankRelevancyWithFeedback":
```

Comment:
@Purg I think I'm misunderstanding something about how this specific section is supposed to work. Do you think you could point out what I'm doing wrong here?
This is the mypy error for this file
```
tests/algorithms/rank_relevancy/test_RR_abstract.py:19: error: Signature of "_rank_with_feedback" incompatible with supertype "RankRelevancyWithFeedback"
tests/algorithms/rank_relevancy/test_RR_abstract.py:40: error: "RankRelevancyWithFeedback" has no attribute "assert_not_called"
tests/algorithms/rank_relevancy/test_RR_abstract.py:56: error: "RankRelevancyWithFeedback" has no attribute "assert_called_once_with"
tests/algorithms/rank_relevancy/test_RR_abstract.py:57: error: "RankRelevancyWithFeedback" has no attribute "return_value"
```

Reply:
I think what @neal-siekierski (mentioning you in case you have 2-cents) may have done the first time maybe doesn't jive so well with the level of additional type annotation checking we are doing here. E.g. this override does not match the signature and return type of the parent method, so that itself is a mypy error. We can revise this by simply defining an empty implementation of this abstract method, and then per-test assigning a `mock.Mock()` instance to the instance method property where applicable. E.g. in `test_rrwf_length_check()`:
```python
def test_rrwf_length_check() -> None:
    """ ... """
    rrwf = DummyRRWF()
    rrwf._rank_with_feedback = mock.Mock()  # type: ignore
    ...
```

It's definitely a less fancy, more-typing (just a little) way of doing this, but it's much simpler to understand and should be more explicit in what is happening.

Reply:
@bardkw, if you're wondering how the code currently works, `rrwf._rank_with_feedback` is normally a bound method when `isinstance(rrwf, RankRelevancyWithFeedback)`. The definition here makes it a `mock.Mock` instead, and that works out since it's still callable like the interface demands. (I don't know if that's terribly helpful since I've largely just repeated the comment.)

I'm not surprised that this is hard to type in mypy, assuming it's even possible. My take would be to make `_rank_with_feedback` a normal method with its normal signature that calls `self._rwf_mock` and then replace uses of `rrwf._rank_with_feedback.return_value` with `rrwf._rwf_mock.return_value` etc.  That avoids overwriting the method at every use site.

(In general, trying to type-check code that uses `mock.Mock` sounds like trouble.)

Reply:
@neal-siekierski when you say a 'normal method with its normal signature', what exactly do you mean? I'm assuming you mean 1)not as a property and 2)with the arguments established in the original method

Reply:
If I understand correctly, @neal-siekierski is saying to restore the method with the interface class defined signature (so yes to [2]), changing the body of the function to something like `return self._rwf_mock(...)` where `...` is the parameter inputs to `_rank_with_feedback`.

Reply:
@bardkw, yes, that's what I had in mind. Thanks for elaborating, @Purg!

Reply:
so this as I understand?
`    @mock.create_autospec
    def _rank_with_feedback(self,
            pos: Sequence[numpy.ndarray],
            neg: Sequence[numpy.ndarray],
            pool: Sequence[numpy.ndarray],
            pool_uids: Sequence[Hashable],) -> Tuple[Sequence[float], Sequence[Hashable]]:
        return self._rwf_mock(self, pos, neg, pool, pool_uids)`